### PR TITLE
NIAD-1947: Check database for snomed immunization code

### PR DESCRIPTION
**Pull request by ElliotB-96**
Started implementation of checking immunization codes from the database

### CODE REVIEW

File: db-connector/src/main/java/uk/nhs/adaptors/connector/dao/ImmunizationSnomedCTDao.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+import org.jdbi.v3.sqlobject.locator.UseClasspathSqlLocator;
+import org.jdbi.v3.sqlobject.statement.SqlQuery;
+
+import uk.nhs.adaptors.connector.model.ImmunizationSnomedCT;
+
+public interface ImmunizationSnomedCTDao {
+    @SqlQuery("select_immunization_concept_id")
+    @UseClasspathSqlLocator
+    ImmunizationSnomedCT getImmunizationSnomednUsingConceptId(@Bind("conceptId") String conceptId);
+}
```
Comment (mrogowska) [inline db-connector/src/main/java/uk/nhs/adaptors/connector/dao/ImmunizationSnomedCTDao.java:13]: Add empty line at the end of the file please

File: db-connector/src/main/java/uk/nhs/adaptors/connector/model/ImmunizationSnomedCT.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+import lombok.Builder;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+@Builder
+public class ImmunizationSnomedCT {
+    private String conceptid;
```
Comment (mrogowska) [inline db-connector/src/main/java/uk/nhs/adaptors/connector/model/ImmunizationSnomedCT.java:11]: Maybe concept**Id** would be better

File: db-connector/src/main/resources/uk/nhs/adaptors/connector/dao/ImmunizationSnomedCTDao/select_immunization_concept_id.sql [hunk 3]
```diff
@@ -0,0 +1 @@
+SELECT d.conceptid, d.description, d.safetycode FROM "snomedct".immunization_codes d WHERE conceptid = :conceptId;
```
Comment (mrogowska) [inline db-connector/src/main/resources/uk/nhs/adaptors/connector/dao/ImmunizationSnomedCTDao/select_immunization_concept_id.sql:1]: Just out of curiosity, why `immunization_codes` table has the alias "d" :D?

File: docker/start-local-environment.sh [hunk 4]
```diff
@@ -11,7 +11,7 @@ docker-compose up -d ps_db mhs-adaptor-mock activemq;
 
 cd ../db-connector
 ./gradlew update
-cd ../docker
+ cd ../docker
```
Comment (mrogowska) [inline docker/start-local-environment.sh:14]: Are those spaces here and before docker-compose commands necessary?

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/util/ImmunizationChecker.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import lombok.RequiredArgsConstructor;
+import uk.nhs.adaptors.connector.dao.ImmunizationSnomedCTDao;
+
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+@Component
+public class ImmunizationChecker implements ImmunizationCheckerInterface {
```
Comment (dawidskainos) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/util/ImmunizationChecker.java:11]: I'd work on the names:
I'd call the interface: ImmunisationChecker and it's implementation - DatabaseImmunisationChecker.

If we call the implementation ImmunisationChecker then it doesn't really say how it's being implemented and doesn't leave too much space for other implementations

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java [hunk 6]
```diff
@@ -68,6 +69,7 @@
 
     private final CodeableConceptMapper codeableConceptMapper;
     private final ConsultationListMapper consultationListMapper;
+    private final ImmunizationChecker immunizationChecker;
```
Comment (mrogowska) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java:72]: I would change the ResourceReferenceUtil methods so they're no longer static and inject ImmunizationChecker there, not here.

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/TemplateMapper.java [hunk 7]
```diff
@@ -43,6 +44,7 @@
     private static final String QUESTIONNAIRE_REFERENCE = "%s-QRSP";
 
     private final CodeableConceptMapper codeableConceptMapper;
+    private final ImmunizationChecker immunizationChecker;
```
Comment (mrogowska) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/TemplateMapper.java:47]: It won't be needed here if it was injected in the ResourceReferenceUtil :D

File: snomed-immunization-loader/README.md [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+- PostgreSQL v.9
+
+## How to run (Mac & Unix)
+
+### Prerequsistes
+1. Format codes into CSV with 3 columns (Concept Id,Description,Safety Code)
+2. Set the follwing env vars:
+- DB_OWNER_USERNAME (database user, needs permissions to create/drop schemas), e.g. `export DB_OWNER_USERNAME='postgres'`
+- PGPASSWORD (database user password), e.g. `export PGPASSWORD='********'`
```
Comment (mrogowska) [inline snomed-immunization-loader/README.md:12]: Should be POSTGRES_PASSWORD

File: snomed-database-loader/create-database-postgres.sql [hunk 9]
```diff
@@ -30,5 +30,11 @@ create table langrefset_s(
 CREATE INDEX langrefset_referencedcomponentid_idx ON snomedct.langrefset_s
     USING btree (referencedcomponentid);
 
+create table immunization_codes(
```
Comment (mrogowska) [inline snomed-database-loader/create-database-postgres.sql:33]: Could you change `create table` to capital letters for consistency?

**Review by mrogowska — CHANGES_REQUESTED** [review]

File: docker/start-local-environment.sh (hunk 4 above)
Comment (ElliotB-96) [reply docker/start-local-environment.sh:14]: Artefact of me tweaking them, removed

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/util/ResourceReferenceUtil.java [hunk 10]
```diff
@@ -99,9 +107,9 @@ private static void addTemplateEntry(RCMRMT030101UK04CompoundStatement compoundS
             compoundStatement.getId().get(0).getRoot()));
     }
 
-    private static void addObservationStatementEntry(RCMRMT030101UK04ObservationStatement observationStatement,
+    private void addObservationStatementEntry(RCMRMT030101UK04ObservationStatement observationStatement,
         List<Reference> entryReferences, RCMRMT030101UK04CompoundStatement compoundStatement,
-        ImmunizationChecker immunizationChecker) {
+        DatabaseImmunizationChecker immunizationChecker) {
```
Comment (mrogowska) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/util/ResourceReferenceUtil.java:112]: Do you need to pass immunizationChecker as a param here?

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapperTest.java [hunk 11]
```diff
@@ -97,18 +97,19 @@
     private ConsultationListMapper consultationListMapper;
 
     @Mock
-    private ImmunizationChecker immunizationChecker;
+    private DatabaseImmunizationChecker immunizationChecker;
 
-    @InjectMocks
```
Comment (mrogowska) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapperTest.java:96]: You could have `@InjectMocks` over EncounterMapper, and ResourceReferenceUtil as a mock

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/TemplateMapperTest.java [hunk 12]
```diff
@@ -52,11 +53,17 @@
     private CodeableConceptMapper codeableConceptMapper;
 
     @Mock
-    private ImmunizationChecker immunizationChecker;
+    private DatabaseImmunizationChecker immunizationChecker;
 
-    @InjectMocks
+    private ResourceReferenceUtil resourceReferenceUtil;
     private TemplateMapper templateMapper;
```
Comment (mrogowska) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/TemplateMapperTest.java:59]: The same comment as for EncounterMapper test, you could now mock ResourceReferenceUtil (then DatabaseImmunizationChecker mock won't be needed)

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/ResourceReferenceUtilTest.java [hunk 13]
```diff
@@ -35,14 +35,14 @@
     private static final int FOUR = 4;
 
     @Mock
-    private ImmunizationChecker immunizationChecker;
+    private DatabaseImmunizationChecker immunizationChecker;
```
Comment (mrogowska) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/ResourceReferenceUtilTest.java:39]: I think you could have `@InjectMocks ResourceReferenceUtil` declaration and then wouldn't need to create the ResourceReferenceUtil object in each test.

**Review by mrogowska — CHANGES_REQUESTED** [review]

**Review by mrogowska — APPROVED** [review]

**Review by rorysproule2 — DISMISSED** [review]

File: snomed-database-loader/create-database-postgres.sql [hunk 14]
```diff
@@ -30,5 +30,11 @@ create table langrefset_s(
 CREATE INDEX langrefset_referencedcomponentid_idx ON snomedct.langrefset_s
     USING btree (referencedcomponentid);
 
+CREATE TABLE immunization_codes(
+                              conceptid varchar(18) not null,
```
Comment (boydstep) [inline snomed-database-loader/create-database-postgres.sql:34]: Are these codes unique?

**Review by boydstep — APPROVED** [review]

**Review by mrogowska — APPROVED** [review]

**Review by dawidskainos — APPROVED** [review]

**Pull request merged**